src/desktopentry.cpp

```diff
@@ -0,0 +1,193 @@
+// SPDX-FileCopyrightText: 2023 UnionTech Software Technology Co., Ltd.
+//
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+#include "desktopentry.h"
+#include <QFileInfo>
+#include <QDir>
+#include <QStandardPaths>
+#include <QTextStream>
+#include <algorithm>
+#include <QRegularExpression>
+#include <QDirIterator>
+
+auto DesktopEntry::parserGroupHeader(const QString &str) noexcept
+{
+    auto groupHeader = str.sliced(1, str.size() - 2);
+    if (auto it = entryMap.find(groupHeader); it == entryMap.cend()) {
+        return entryMap.insert(groupHeader, {});
+    } else {
+        return it;
+    }
+}
+
+DesktopEntry::ParseError DesktopEntry::parseEntry(const QString &str, decltype(entryMap)::iterator& currentGroup) noexcept
+{
+    if (str.startsWith("#"))
+        return ParseError::NoError;
+    auto splitCharIndex = str.indexOf("=");
+    auto keyStr = str.first(splitCharIndex).trimmed();
+    auto valueStr = str.sliced(splitCharIndex + 1).trimmed();
+    QString key, valueKey{defaultKeyStr};
+    
+    if (keyStr.contains(QRegularExpression{"[^0-9a-zA-Z-]"})) {
+        if(keyStr.count('[') == 1 and keyStr.count(']') == 1 and keyStr.back() == ']'){
+            auto tmp = keyStr.split('[');
+            key = tmp[0];
+            valueKey = tmp[1].chopped(1);
+        } else {
+            qWarning() << "invalid key: " << keyStr;
+            return ParseError::EntryKeyInvalid;
+        }
+    }
+
+    if (auto cur = currentGroup->find(key); cur == currentGroup->end()) {
+        currentGroup->insert(keyStr, {{valueKey, valueStr}});
+    } else {
+        if (auto v = cur->find(valueKey); v == cur->end()) {
+            cur->insert(valueKey, valueStr);
+        }
+    }
+    return ParseError::NoError;
+}
+
+DesktopEntry::ParseError DesktopEntry::parse(const QString &desktopFileName) noexcept
```

Comment:
> Desktop entry files should have the .desktop extension, except for files of Type Directory which should have the .directory extension. Determining file type on basis of extension makes determining the file type very easy and quick. When no file extension is present, the desktop system should fall back to recognition via "magic detection".

I think we should refer to the spec, or just copy the spec into comments, when we try to decide something is invalid.